### PR TITLE
Examples gallery for napari-threedee

### PR DESCRIPTION
**Pull request by GenevieveBuckley**
Closes https://github.com/napari-threedee/napari-threedee/issues/99

image

### CODE REVIEW

File: mkdocs.yml [hunk 1]
```diff
@@ -130,6 +131,21 @@ plugins:
...
+        docs/examples/library,
+        docs/examples/plugin
+      ]
+      gallery_dirs: [
+        docs/generated/gallery,
+        docs/generated/gallery/library,
+        docs/generated/gallery/plugin
+      ]
+      filename_pattern: /*.py  # which scripts will be executed for the docs
+      ignore_pattern: /mesh_headlight*
```
Comment (GenevieveBuckley) [inline mkdocs.yml:147]: This line can be removed when issue https://github.com/napari-threedee/napari-threedee/issues/103 is resolved

**Comment by GenevieveBuckley** [comment]
Here's what it looks like currently:
Screen Shot 2023-03-24 at 5 04 06 pm

**Comment by kevinyamauchi** [comment]
> Here's what it looks like currently: Screen Shot 2023-03-24 at 5 04 06 pm

Awesome! This looks great!

I see that there are multiple "gallery subpages" in the nav. Would it be possible to make a separate subpage for the "library" examples and the "plugin examples"?

I think it isn't necessary for the initial PR, but at some point, it might be nice to see if it's possible to make the screenshot have some of the manipulators, annotators, etc. in view.

Super cool stuff, @GenevieveBuckley !

**Comment by kevinyamauchi** [comment]
I just saw https://github.com/napari-threedee/napari-threedee/issues/103. I can work on that today.

File: setup.cfg [hunk 2]
```diff
@@ -52,11 +52,13 @@ napari.manifest =
 [options.extras_require]
 dev = 
 	mkdocs
```
Comment (GenevieveBuckley) [inline setup.cfg:54]: We'll have to pin to mkdocs version 1.2.4 for the moment if we want mkdocs-gallery to work, see https://github.com/smarie/mkdocs-gallery/issues/57

**Comment by GenevieveBuckley** [comment]
Here's the latest gallery screenshot:

image

File: setup.cfg (hunk 2 above)
Comment (GenevieveBuckley) [reply setup.cfg]: Pinned

**Comment by GenevieveBuckley** [comment]
@kevinyamauchi 

> I see that there are multiple "gallery subpages" in the nav. Would it be possible to make a separate subpage for the "library" examples and the "plugin examples"? You may have already inferred, but former shows how to import the actual widget and the latter shows how to use the plugin interface.

Yes, this is already done (sorry, I just named things poorly in the last screenshot, so it looked like duplicated gallery subpages, instead of being named for the plugin and library examples)

> I think it isn't necessary for the initial PR, but at some point, it might be nice to see if it's possible to make the screenshot have some of the manipulators, annotators, etc. in view.

Yes, good idea for the future.

**Comment by GenevieveBuckley** [comment]
One slightly weird thing to be aware of: I had to turn the screenshot flash animation OFF in napari_image_scraper, because otherwise it would turn some of the image scheenshots grey somehow. 

I also don't understand your example script organisation. At first it looks like the examples are just duplicated across the "plugin" and "library" folders, but actually not all scripts are the same, and sometimes even the ones that are named the same have slightly different code and/or comments.

**Comment by GenevieveBuckley** [comment]
Ok, I don't know how to initialize Qt properly on the CI.

```
WARNING: Could not load the Qt platform plugin "xcb" in "" even though it was found.
WARNING: This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
```

I though adding Talley's github action might fix it, but no. How does the main napari repo do this?

**Comment by GenevieveBuckley** [comment]
Hooray! Andy and Draga helped me figure out the github actions workflow.

We needed:
- Talley's setup qt github action (for all the dependencies you need to have before qt is installed with napari later) https://github.com/tlambert03/setup-qt-libs
- Ashley's headless GUI github action (so there is a virtual display for the CI) https://github.com/aganders3/headless-gui
- ...and finally we had to fix the syntax of the workflow text document. It's not that intuitive, if you have `uses` in a step, then you need to put `run` inside a with block, etc. https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions

**Review by kevinyamauchi — COMMENTED** [review]
Wow, this is awesome! 

it looks like the `--strict` mkdocs build is failing because:
- our filenames for the `library` and `plugin` examples are the same. I can fix this in a separate PR
- there are references to non-existant API docs in the nav config. I can fix this in a separate PR.

I'm not familiar with the merits of building with/without the `--strict` flag. Do you think we should use it, @GenevieveBuckley ? If so, I can make the PRs I mentioned above and I think it will work :)

**Comment by kevinyamauchi** [comment]
> I also don't understand your example script organisation. At first it looks like the examples are just duplicated across the "plugin" and "library" folders, but actually not all scripts are the same, and sometimes even the ones that are named the same have slightly different code and/or comments.

Sorry, that's a bit sloppy on our side. The intention is that we have examples that both import the actual widget and launch with the plugin. The reason to show the widget import is just so folks that want to make custom widgets by composing our widgets with other know where to start. We can make sure the code is consistent in a separate PR.

**Comment by GenevieveBuckley** [comment]
Sorry - forgot to edit the title, I'll take out the "WIP" bit now

**Review by kevinyamauchi — APPROVED** [review]
Wow, this is super cool, @GenevieveBuckley . Thank you for contributing this and for figuring out how to get it all running on CI.

I tested building it locally and it looks great. I kind of like having the examples scripts in the base repo so that they are easy for folks checking out the repo to find. I _think_ we could adjust the CI job to copy the files from the `examples` directory into the `docs` at runtime. However, I think this PR is a standalone contribution as it establishes the infrastructure and we can iterate in future PRs.

I will leave this up for the rest of the day in case @alisterburt wants to weigh in. Otherwise, I will merge tomorrow. Thanks again!

**Pull request merged**